// angular2-hotkeys/esm2015/lib/hotkey.module.js
var HotkeyModule_1;
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { HotkeysDirective } from './hotkeys.directive';
import { HotkeysCheatsheetComponent } from './hotkeys-cheatsheet/hotkeys-cheatsheet.component';
import { CommonModule } from '@angular/common';
import { HotkeyOptions } from './hotkey.options';
import { HotkeysService } from './hotkeys.service';
let HotkeyModule = HotkeyModule_1 = class HotkeyModule {
    // noinspection JSUnusedGlobalSymbols
    static forRoot(options = {}) {
        return {
            ngModule: HotkeyModule_1,
            providers: [
                HotkeysService,
                { provide: HotkeyOptions, useValue: options }
            ]
        };
    }
};
HotkeyModule = HotkeyModule_1 = __decorate([
    NgModule({
        declarations: [HotkeysDirective, HotkeysCheatsheetComponent],
        imports: [CommonModule],
        exports: [HotkeysDirective, HotkeysCheatsheetComponent]
    })
], HotkeyModule);
export { HotkeyModule };
