// angular2-hotkeys/esm5/lib/hotkey.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { HotkeysDirective } from './hotkeys.directive';
import { HotkeysCheatsheetComponent } from './hotkeys-cheatsheet/hotkeys-cheatsheet.component';
import { CommonModule } from '@angular/common';
import { HotkeyOptions } from './hotkey.options';
import { HotkeysService } from './hotkeys.service';
var HotkeyModule = /** @class */ (function () {
    function HotkeyModule() {
    }
    HotkeyModule_1 = HotkeyModule;
    // noinspection JSUnusedGlobalSymbols
    HotkeyModule.forRoot = function (options) {
        if (options === void 0) { options = {}; }
        return {
            ngModule: HotkeyModule_1,
            providers: [
                HotkeysService,
                { provide: HotkeyOptions, useValue: options }
            ]
        };
    };
    var HotkeyModule_1;
    HotkeyModule = HotkeyModule_1 = __decorate([
        NgModule({
            declarations: [HotkeysDirective, HotkeysCheatsheetComponent],
            imports: [CommonModule],
            exports: [HotkeysDirective, HotkeysCheatsheetComponent]
        })
    ], HotkeyModule);
    return HotkeyModule;
}());
export { HotkeyModule };
