// @angular-material-components/moment-adapter/esm5/lib/moment-adapter.module.js
/**
 * @fileoverview added by tsickle
 * Generated from: lib/moment-adapter.module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { NgxMatDateAdapter, NGX_MAT_DATE_FORMATS } from '@angular-material-components/datetime-picker';
import { NgModule } from '@angular/core';
import { MAT_DATE_LOCALE } from '@angular/material/core';
import { NgxMatMomentAdapter, NGX_MAT_MOMENT_DATE_ADAPTER_OPTIONS } from './moment-adapter';
import { NGX_MAT_MOMENT_FORMATS } from './moment-formats';
var NgxMomentDateModule = /** @class */ (function () {
    function NgxMomentDateModule() {
    }
    NgxMomentDateModule.decorators = [
        { type: NgModule, args: [{
                    providers: [
                        {
                            provide: NgxMatDateAdapter,
                            useClass: NgxMatMomentAdapter,
                            deps: [MAT_DATE_LOCALE, NGX_MAT_MOMENT_DATE_ADAPTER_OPTIONS]
                        }
                    ],
                },] }
    ];
    return NgxMomentDateModule;
}());
export { NgxMomentDateModule };
var ɵ0 = NGX_MAT_MOMENT_FORMATS;
var NgxMatMomentModule = /** @class */ (function () {
    function NgxMatMomentModule() {
    }
    NgxMatMomentModule.decorators = [
        { type: NgModule, args: [{
                    imports: [NgxMomentDateModule],
                    providers: [{ provide: NGX_MAT_DATE_FORMATS, useValue: ɵ0 }],
                },] }
    ];
    return NgxMatMomentModule;
}());
export { NgxMatMomentModule };
export { ɵ0 };
